src/codegeneration/ES6ClassTransformer.js

```diff
@@ -0,0 +1,259 @@
+            constructor = tree;
+            constructorIndex = elements.length;
+          } else {
+            elements.push(this.transformAny(tree));
+          }
+          break;
+
+        case PROPERTY_VARIABLE_DECLARATION:
+          tree = this.transformAny(tree);
+          if (tree.initializer !== null) {
+            initVars.push(tree);
+          }
+          break;
+
+        default:
+          throw new Error(`Unexpected class element: ${tree.type}`);
+      }
+    });
+
+    if (initInstanceVars.length > 0) {
+      let initExpression = getInstanceInitExpression(initInstanceVars);
+
+      if (!constructor) {
+        constructor = this.getDefaultConstructor_(!!tree.superClass);
+      }
+
+      constructor = this.appendInstanceInitializers_(constructor,
+          initExpression, tree.superClass);
+    }
+
+    if (constructor) {
+      elements.splice(constructorIndex, 0, constructor);
+    }
+
+    return {
+      elements,
+      initStaticVars,
+    };
+  }
+
+  /**
+   * Transforms a single class declaration
+   *
+   * @param {ClassDeclaration} tree
+   * @return {ParseTree}
+   */
+  transformClassDeclaration(tree) {
+    let {
+      elements,
+      initStaticVars,
+    } = this.transformClassElements_(tree);
+
+    let statements = createStaticInitializerStatements(tree.name, initStaticVars);
+    let superClass = this.transformAny(tree.superClass);
+    let classDecl = new ClassDeclaration(tree.location, tree.name, superClass,
+        elements, tree.annotations, tree.typeParameters);
+
+    statements = prependStatements(statements, classDecl);
+
+    return new AnonBlock(null, statements);
+  }
+
+  /**
+   * Transforms a single class expression
+   *
+   * @param {ClassExpression} tree
+   * @return {ParseTree}
+   */
+  transformClassExpression(tree) {
+    let {
+      elements,
+      initStaticVars,
+    } = this.transformClassElements_(tree);
+
+    let superClass = this.transformAny(tree.superClass);
+    let classExpression = new ClassExpression(tree.location, tree.name,
+        superClass, elements, tree.annotations, tree.typeParameters);
+
+    if (initStaticVars.length === 0) {
+      return classExpression;
+    }
+
+    this.pushTempScope();
+    let id = this.getTempIdentifier();
+    let className = createBindingIdentifier(id);
+    let statements = [
+      parseStatement `let ${id} = ${classExpression}`,
+      ...createStaticInitializerStatements(className, initStaticVars),
+      parseStatement `return ${className}`
+    ];
+    let body = createFunctionBody(statements);
+    this.popTempScope();
+
+    return createImmediatelyInvokedFunctionExpression(body);
+  }
+
+  getDefaultConstructor_(hasSuper) {
+    if (hasSuper) {
+      var pma = parsePropertyDefinition `constructor(...args) {}`;
+      // `parsePropertyDefinition()` would not allow `super(...)` in method body
+      pma.body.statements = [parseStatement `super(...args)`];
```

Comment:
ParseTrees are immutable by a gentlemen's agreement. By mutating it the caching will break.

I think for now you need to create a new tree (or fix parsePropertyDefinition).